### PR TITLE
feat: web streams based archive/tar

### LINKED ISSUE #1658: Modernize `std/archive/tar`
The current `std/archive/tar` builds on `Deno.Reader` and friends. We should rewrite it to use WHATWG streams. So a `TarDecoderStream` and `TarEncoderStream`.

### PR DESCRIPTION
**Pull request by crowlKats**
still WIP

closes #1658

### CODE REVIEW

File: archive/tar_stream_test.ts [hunk 1]
```diff
@@ -0,0 +1,499 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
```
Comment (jsejcksn) [inline archive/tar_stream_test.ts:1]:
```suggestion
// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
```

File: archive/tar_stream.ts [hunk 2]
```diff
@@ -0,0 +1,549 @@
+/**
+ * Ported and modified from: https://github.com/beatgammit/tar-js and
+ * licensed as:
+ *
+ * (The MIT License)
+ *
+ * Copyright (c) 2011 T. Jameson Little
+ * Copyright (c) 2019 Jun Kato
+ * Copyright (c) 2018-2021 the Deno authors
```
Comment (jsejcksn) [inline archive/tar_stream.ts:9]:
```suggestion
 * Copyright (c) 2018-2022 the Deno authors
```

**Comment by bartlomieju** [comment]
@crowlKats there's already https://github.com/denoland/deno_std/pull/1855. Should one of the PRs be closed?

**Comment by crowlKats** [comment]
@bartlomieju well, the approach of the other PR wont work, and this PR right now is stuck due to upstream spec change needed

**Comment by danopia** [comment]
> due to upstream spec change needed

I'm curious if there is a reference on what is needed / where the conversation is?

Also as an outsider it seems that the `filePath` shorthand parameter will not be available in the new API to allow removing all `Deno.*` dependencies, is that right?

Cross link:
* #1658

**Comment by bartlomieju** [comment]
Still blocked by spec update

**Comment by iuioiua** [comment]
@crowlKats, are there any possible workarounds for this PR? Perhaps, some alternative API that's come up since the last update or a WASM alternative? Disclaimer: I'm unfamiliar with this implementation.

**Comment by crowlKats** [comment]
@iuioiua I sadly believe no

**Comment by csvn** [comment]
The https://github.com/whatwg/streams/pull/1145 PR has now been merged. Is this PR now unblocked, or does this require V8/Deno updates before STD can be updated?

**Comment by csvn** [comment]
Nevermind, sorry. Saw there is work on this in https://github.com/denoland/deno/pull/20849.

**Comment by crowlKats** [comment]
yep, once that PR lands in Deno, I'll work on this PR again

**Comment by iuioiua** [comment]
Update: the [`deflateTarArchive` test](https://github.com/crowlKats/deno_std/blob/c9913e8e75fc81db86ac4bc70fdeada1887596a0/archive/untar_stream_test.ts#L57-L92) is failing because the `TarEntry.#readableInner` is locked when executing [this line](https://github.com/crowlKats/deno_std/blob/c9913e8e75fc81db86ac4bc70fdeada1887596a0/archive/untar_stream.ts#L105). Any insight would be welcome.

File: streams/buffer.ts
Comment (crowlKats) [inline streams/buffer.ts:1]: I'll open a separate PR for these changes

File: archive/untar_stream_test.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    },
+  ];
+
+  const tar = ReadableStream.from<TarOptions>(entries).pipeThrough(new TarStream());
+
+  const untar = new UntarStream();
+  // read data from a tar archive
+  await tar.pipeTo(untar.writable);
+
+  for await (const entry of untar.readable) {
```
Comment (crowlKats) [inline archive/untar_stream_test.ts:230]: when replacing this with a pipeThrough call instead, it breaks. needs to be figured out.

File: archive/untar_stream_test.ts [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    const expected = entries.shift();
+    assert(expected);
+
+    const buffer = new Buffer();
+    await entry.readable.pipeTo(buffer.writable);
+    assertEquals(
+      await toArrayBuffer(expected.readable),
+      await toArrayBuffer(buffer.readable),
+    );
```
Comment (iuioiua) [inline archive/untar_stream_test.ts:98]: Circular test... My mistake.

File: streams/buffer.ts
Comment (crowlKats) [reply streams/buffer.ts:1]: #3982

**Comment by crowlKats** [comment]
Blocked on #3982

**Comment by iuioiua** [comment]
@crowlKats, are you able to complete documentation so `deno doc --lint archive/tar_stream.ts` and `deno doc --lint archive/untar_stream.ts` pass?

**Comment by 91pornaaaaa** [comment]
https://91porna.com/

**Pull request closed without merging**